### PR TITLE
[2091][performance] Track throughput metrics

### LINKED ISSUE #2091: Performance metrics
### Describe the task. It can be a feature, documentation, etc.

Metrics to evaluate our model's computational performance on.

Requirements:
- comparable across datasets (normalize by token size or similar)
- normalized by time.

The goal is to in to integrate these into performance tests that we run on a regular basis (see https://github.com/ecmwf/WeatherGenerator/issues/2092) and track these in MLFlow (see https://github.com/ecmwf/WeatherGenerator/issues/1930).

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

https://gitlab.jsc.fz-juelich.de/hedgedoc/KCeuWSa3S86yc2EmUUtzyQ?both#

### Area

- [ ] datasets, data readers, data preparation and transfer
- [ ] model
- [ ] science
- [x] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation

### PR DESCRIPTION
**Pull request by florianscheidl**
## Description
Implements optional on-the-fly throughput metrics, logged per training step. The new metrics are named "performance.throughput.*" and track per-device and global throughput in terms of:
- batches per second,
- samples per second,
- MB per second.

In multi-device and multi-node setups, we make an all-reduce call to get the global throughput metrics.

### Usage
To activate throughput tracking, add `track_performance_metrics: True` in the training config, under `train_logging`, see the performance_*.yaml configs added here. Run with a base configuration, e.g.:
```
../WeatherGenerator-private/hpc/launch-slurm.py --time 8 --nodes=1 --base-config config/config_jepa.yml --config config/performance_jepa_config.yml
```

## Issue Number

Closes #2091.

## Preview:

We investigated the effect of batch sizes on throughput, see https://gitlab.jsc.fz-juelich.de/hedgedoc/SUW6Zq-BR3uYCwU3hmIb6w?both#.

Below are screenshots from MLFlow:

Screenshot 2026-04-02 at 17 48 06

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -83,6 +89,16 @@ def __init__(self, train_logging: Config):
         self.batch_size_validation_per_gpu = -1
         self.batch_size_test_per_gpu = -1
         self.collapse_monitor: CollapseMonitor | None = None
+        self.throughput: Throughput | None = None
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Please avoid having many more variables and code in Trainer. It's a critical and complex part in the code that we need to keep as clean as possible.

These variables could go into a small separate class.

File: src/weathergen/train/trainer.py [hunk 2]
```diff
@@ -158,6 +174,23 @@ def init(self, cf: Config, devices):
         collapse_config = cf.train_logging.get("collapse_monitoring", {})
         self.collapse_monitor = CollapseMonitor(collapse_config, None)  # device set later in run()
 
+        if cf.train_logging.get("track_performance_metrics"):
```
Comment (clessig) [inline src/weathergen/train/trainer.py:164]: Can this be implemented in the class above?

File: src/weathergen/train/trainer.py [hunk 3]
```diff
@@ -367,6 +400,15 @@ def run(self, cf, devices, run_id_contd=None, mini_epoch_contd=None):
             config.save(self.cf, None)
             logger.info(config.format_cf(self.cf))
 
+        if self.cf.train_logging.get("track_performance_metrics"):
+            # precompute batch statistics for throughput/MFU tracking
+            sample_batch = next(iter(self.data_loader))
+            sample_batch.to_device(self.device)
+            self.precompute_flops(sample_batch)
+            self.precompute_source_bytes(sample_batch)
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Note this is not a constant but might fluctuate widely between samples.

File: src/weathergen/train/trainer.py [hunk 4]
```diff
@@ -367,6 +400,15 @@ def run(self, cf, devices, run_id_contd=None, mini_epoch_contd=None):
             config.save(self.cf, None)
             logger.info(config.format_cf(self.cf))
 
+        if self.cf.train_logging.get("track_performance_metrics"):
+            # precompute batch statistics for throughput/MFU tracking
+            sample_batch = next(iter(self.data_loader))
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: With this we change the training samples and logic--can't we collect on the fly in train, perhaps sparsely if there are performance concerns (see also the comment right below).

File: src/weathergen/train/trainer.py [hunk 5]
```diff
@@ -410,6 +452,65 @@ def validate_before_training(self):
             else:
                 assert False, "validate_before_training must be integer or boolean."
 
+    def _compute_targets_and_auxs(self, sample_batch) -> dict:
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Why has this changed/moved?

File: src/weathergen/train/trainer.py [hunk 6]
```diff
@@ -410,6 +452,65 @@ def validate_before_training(self):
...
+            target_idxs = get_target_idxs_from_cfg(self.training_cfg, loss_name)
+            targets_and_auxs[loss_name] = target_aux.compute(
+                self.cf.general.istep,
+                sample_batch.get_target_samples(target_idxs),
+                self.model_params,
+                self.model,
+            )
+        return targets_and_auxs
+
+    def precompute_flops(self, sample_batch) -> None:
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: I really have concerns that we run training before we start training. Eg, for model dev any bug will be triggered here and not in the actual training loop--this is not feasible.

File: src/weathergen/train/trainer.py [hunk 7]
```diff
@@ -529,6 +647,22 @@ def train(self, mini_epoch):
             self._log_terminal(bidx, mini_epoch, TRAIN)
             if bidx % self.train_logging.metrics == 0:
                 self._log(TRAIN)
+                # Log throughput metrics
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Again, too much code here

File: src/weathergen/train/trainer.py [hunk 8]
```diff
@@ -516,6 +617,23 @@ def train(self, mini_epoch):
             if self.validate_with_ema:
                 self.ema_model.update(self.cf.general.istep * batch_size_total, batch_size_total)
 
+            if self.throughput:
+                torch.cuda.synchronize()
+                if not self._throughput_warmup_done:
+                    if self.cf.general.istep == self._THROUGHPUT_WARMUP_STEPS - 1:
+                        self._t0_throughput = time.time()
+                        self._throughput_warmup_done = True
+                else:
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: What is happening here exactly?

File: src/weathergen/train/trainer.py (hunk 2 above)
Comment (florianscheidl) [reply src/weathergen/train/trainer.py:164]: Refactored code into class init method.

File: src/weathergen/train/trainer.py (hunk 7 above)
Comment (florianscheidl) [reply src/weathergen/train/trainer.py]: Refactored into a single call.

File: tests/test_performance_utils.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+# ---------------------------------------------------------------------------
+# ThroughputTracker
+# ---------------------------------------------------------------------------
+
+
+@pytest.fixture()
+def tracker():
+    """A tracker with warmup_steps=2 on CPU."""
+    return ThroughputTracker(device=torch.device("cpu"), world_size=1, warmup_steps=2)
```
Comment (ekouts) [inline tests/test_performance_utils.py]:
```suggestion
    return ThroughputTracker(device=torch.device("cpu"), warmup_steps=2)
```
The signature of this is wrong, right? `world_size` doesn't exist in the `__init__` of the `ThroughputTracker` class

File: tests/test_performance_utils.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    tracker.update(batch_size_per_gpu=4, istep=3, source_mb=1.5)
+
+    assert tracker._total_batches == 2
+    assert tracker._total_samples == 8
+    assert tracker._total_mb == pytest.approx(2.5)
+
+
+def test_warmup_steps_not_counted():
+    """Steps during warmup do not contribute to totals."""
+    tracker = ThroughputTracker(device=torch.device("cpu"), world_size=1, warmup_steps=3)
```
Comment (ekouts) [inline tests/test_performance_utils.py]: Same comment as the fixture

**Review by ekouts — CHANGES_REQUESTED** [review]

File: src/weathergen/utils/performance.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        """Record one training step, handling warmup internally.
+
+        Args:
+            batch_size_per_gpu: Number of samples processed on this rank.
+            istep: Global training step index (used for warmup countdown).
+            source_mb: Source tensor megabytes for this batch. Should be computed
+                       fresh each step via ``compute_source_bytes`` as batch sizes
+                       can vary across samples.
+        """
+        torch.cuda.synchronize()
```
Comment (ekouts) [inline src/weathergen/utils/performance.py]: Do we need to synchronize in every step or should we skip the warm up?

**Review by ekouts — CHANGES_REQUESTED** [review]

File: src/weathergen/utils/performance.py (hunk 11 above)
Comment (florianscheidl) [reply src/weathergen/utils/performance.py]: I think we can skip the warmup 👍

File: tests/test_performance_utils.py (hunk 9 above)
Comment (florianscheidl) [reply tests/test_performance_utils.py]: Thx for flagging!

**Review by ekouts — APPROVED** [review]

File: config/config_performance_default.yml [hunk 12]
```diff
@@ -0,0 +1,32 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
```
Comment (clessig) [inline config/config_performance_default.yml:1]: Rename in `config_performance_default`

File: config/config_performance_jepa.yml [hunk 13]
```diff
@@ -0,0 +1,52 @@
+# (C) Copyright 2025 WeatherGenerator contributors.
```
Comment (clessig) [inline config/config_performance_jepa.yml:1]: Rename in `config_performance_jepa`

File: src/weathergen/train/trainer.py [hunk 14]
```diff
@@ -517,6 +525,14 @@ def train(self, mini_epoch):
             if self.validate_with_ema:
                 self.ema_model.update(self.cf.general.istep * batch_size_total, batch_size_total)
 
+            self.perf_tracker.step(
+                batch,
+                self.batch_size_per_gpu,
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Why is this not passed in the constructor?

File: src/weathergen/train/trainer.py [hunk 15]
```diff
@@ -517,6 +525,14 @@ def train(self, mini_epoch):
             if self.validate_with_ema:
                 self.ema_model.update(self.cf.general.istep * batch_size_total, batch_size_total)
 
+            self.perf_tracker.step(
+                batch,
```
Comment (clessig) [inline src/weathergen/train/trainer.py:530]: For what is batch needed?

**Review by clessig — APPROVED** [review]
Some minor points but essentially ready to be merged

File: src/weathergen/train/trainer.py (hunk 15 above)
Comment (florianscheidl) [reply src/weathergen/train/trainer.py:530]: We compute the throughput in MB from the batch.

**Review by clessig — APPROVED** [review]
